[20.05] Switch to lxml for xml parsing

mvdbeek opened the pull request:
In https://github.com/galaxyproject/galaxy/pull/9610 I started parsing tool documents using lxml for better speed. Unfortunately you can't add `xml.etree.ElementTree.Element`s to documents parsed using lxml, so when we add some dynamic parameters like the JobResourceParameters we get this:
```
galaxy.tools.toolbox.base ERROR 2020-05-11 12:41:10,107 Error reading tool from path: toolshed.g2.bx.psu.edu/repos/devteam/bowtie2/749c918495f7/bowtie2/bowtie2_wrapper.xml
Traceback (most recent call last):
  File "/cvmfs/test.galaxyproject.org/galaxy/lib/galaxy/tools/toolbox/base.py", line 657, in _load_tool_tag_set
    tool = self.load_tool(concrete_path, guid=guid, tool_shed_repository=tool_shed_repository, use_cached=False, tool_cache_data_dir=tool_cache_data_dir)
  File "/cvmfs/test.galaxyproject.org/galaxy/lib/galaxy/tools/toolbox/base.py", line 851, in load_tool
    tool = self.create_tool(config_file=config_file, tool_shed_repository=tool_shed_repository, guid=guid, tool_cache_data_dir=tool_cache_data_dir, **kwds)
  File "/cvmfs/test.galaxyproject.org/galaxy/lib/galaxy/tools/__init__.py", line 298, in create_tool
    tool = self._create_tool_from_source(tool_source, config_file=config_file, **kwds)
  File "/cvmfs/test.galaxyproject.org/galaxy/lib/galaxy/tools/__init__.py", line 316, in _create_tool_from_source
    return create_tool_from_source(self.app, tool_source, **kwds)
  File "/cvmfs/test.galaxyproject.org/galaxy/lib/galaxy/tools/__init__.py", line 237, in create_tool_from_source
    tool = ToolClass(config_file, tool_source, app, **kwds)
  File "/cvmfs/test.galaxyproject.org/galaxy/lib/galaxy/tools/__init__.py", line 508, in __init__
    self.populate_resource_parameters(tool_source)
  File "/cvmfs/test.galaxyproject.org/galaxy/lib/galaxy/tools/__init__.py", line 1318, in populate_resource_parameters
    inputs.append(resource_xml)
TypeError: Argument 'element' has incorrect type (expected lxml.etree._Element, got xml.etree.ElementTree.Element)
``` 

Here I switched all of our ElementTree usage to lxml (if available) and I also tried unifying the names with which we do the imports. The central place to import etree from should be `galaxy.util` now, which falls back to stdlib if lxml is not available (as could happen when using `galaxy-util`, `galaxy-tool-util`, etc).

mvdbeek committed:
Switch to lxml for xml parsing

mvdbeek committed:
Fall back to stdlib xml.etree if lxml not available

mvdbeek committed:
Adapt to different error emitted by lxml

mvdbeek committed:
Import etree from galaxy.util everywhere

And prefer calling things etree (instead of ElementTree, XmlET or ET),
so we can exchange this more easily should that be necessary.

mvdbeek committed:
Pass remove_comments to lxml parser

natefoo commented on lib/galaxy/jobs/runners/util/cli/job/torque.py:
```diff
@@ -83,7 +80,7 @@ def parse_status(self, status, job_ids):
         for line in status.strip().splitlines():
             try:
-                tree = et.fromstring(line.strip())
+                tree = parse_xml_string(line.strip)
```
Did you mean to remove the parens here?

mvdbeek replied:
Definitely not, thanks!
```suggestion
                tree = parse_xml_string(line.strip())
```

natefoo commented on lib/galaxy/authnz/__init__.py:
```diff
@@ -24,13 +24,13 @@ def __init__(self, provider, config, backend_config):
         :param provider: is the name of the identity provider (e.g., Google).
 
-        :type config: xml.etree.ElementTree.Element
+        :type config: lxml.etree.ElementTree._Element
```
Just a doc so I'm probably bikeshedding here but does this make sense when the parser might not be lxml?

natefoo commented on lib/galaxy/tool_shed/galaxy_install/tools/tool_panel_manager.py:
```diff
@@ -109,24 +112,24 @@ def generate_tool_elem(self, tool_shed, repository_name, changeset_revision, own
+            tool_elem = etree.SubElement(tool_section, 'tool')
         else:
-            tool_elem = XmlET.Element('tool')
+            tool_elem = etree.Element('tool')
```
Need to get `Element` first with `Element = getattr(etree, '_Element', etree.Element)` as you did elsewhere?

natefoo commented on lib/galaxy/tool_shed/tools/data_table_manager.py:
```diff
@@ -18,21 +18,21 @@ def generate_repository_info_elem(self, tool_shed, repository_name, changeset_re
         """Create and return an ElementTree repository info Element."""
         if parent_elem is None:
-            elem = XmlET.Element('tool_shed_repository')
+            elem = etree.Element('tool_shed_repository')
```
`Element = getattr(etree, '_Element', etree.Element)`?

natefoo commented on lib/galaxy/tools/parameters/__init__.py:
```diff
@@ -273,7 +273,7 @@ def populate_state(request_context, inputs, incoming, state, errors={}, prefix='
     Populates nested state dict from incoming parameter values.
     >>> from collections import OrderedDict
-    >>> from xml.etree.ElementTree import XML
+    >>> from lxml.etree import XML
```
```suggestion
    >>> from galaxy.util import XML
```
?

natefoo commented on lib/galaxy/util/__init__.py:
```diff
@@ -236,34 +242,39 @@ def unique_id(KEY_SIZE=128):
-            if elem.tail is not None:
-                elem.tail = elem.tail.strip()
-    except ParseError:
+        # Restore ENOENT that would otherwise be an OSError in lxml
```
Weird filesystem semantics make it so it's usually safer to try open, catch, and raise if opening fails (usually faster than checking for existence first too) - is the `OSError`'s `errno` attribute set incorrectly in the case of lxml?

Plus IOErrors are OSErrors in Python 3.

mvdbeek replied:
> is the `OSError`'s `errno` attribute set incorrectly in the case of lxml?

indeed, and that leads to this fun undefined behavior where we catch `IOError`s but then don't do anything if errno is not `ENOENT` - https://github.com/mvdbeek/galaxy/blob/a253acc31bf8aaa903e8e5fdbe44b31b204a0393/lib/galaxy/tools/data_manager/manager.py#L71.

I'll add another `except IOError` and change the errno there if appropriate.

mvdbeek replied:
`e.errno` is `None` for missing files. I've added the missing errno in https://github.com/galaxyproject/galaxy/pull/9759/commits/0a8634b2f70be23c3fae85b88d067844acc2b998, I think this should be good enough as a workaround.

natefoo commented on lib/galaxy/util/tool_shed/xml_util.py:
```diff
@@ -47,7 +28,7 @@ def create_element(tag, attributes=None, sub_elements=None):
     """
     if tag:
-        elem = XmlET.Element(tag)
+        elem = etree.Element(tag)
```
`Element = getattr(etree, '_Element', etree.Element)` ?

mvdbeek replied:
This is only necessary for the `isinstance` check, since the actual `Element` class is `_Element` and `Element` is the constructor.

mvdbeek committed:
Also import etree from galaxy.util in doctests

Co-authored-by: Nate Coraor <nate@bx.psu.edu>

mvdbeek committed:
Restore line.strip()

mvdbeek committed:
Add missing errno to exception raised by lxml for missing file

mvdbeek committed:
Fallback to creating XML elements from bytestrings

This is necessary when calling parse_xml_string on (unicode) strings
with an encoding declaration.
